File: 0049 Group Anagrams/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) 
    {
        unordered_map<string, vector<string>> mp;
        for (string s: strs)
        {
            string t = s;
            sort(t.begin(), t.end());
            mp[t].push_back(s);
        }
        
        vector<vector<string>> res;
        for (auto item: mp)
            res.push_back(item.second);
        return res;
    }
};

int main(){
    vector<string> vec{"eat", "tea", "tan", "ate", "nat", "bat"};
    for (auto r : Solution().groupAnagrams(vec))
    {
        for (auto a : r)
            cout << a << " ";
        cout << endl;
    }

    return 0;
}

File: 0049 Group Anagrams/solution.py
#! python3
import collections

class Solution:
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        res = collections.defaultdict(list)
        for s in strs:
            res[tuple(sorted(s))].append(s)
        return list(res.values())

if __name__ == "__main__":
    print(Solution().groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))